src/com/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -883,6 +884,108 @@
     (sql/do-commands
      "DROP INDEX fact_values_string_trgm")))
 
+(defn- lift-fact-paths-into-facts
+  "Pairs paths and values directly in facts, i.e. change facts from (id
+  value) to (id path value)."
+  []
+  (sql/do-commands
+   (if (scf-utils/postgres?)
+     "SET CONSTRAINTS ALL DEFERRED" ;; In case it allows additional optimization?
+     "SELECT 1")
+
+   ;; Build complete facts table as of migration 28.
+
+   "CREATE TABLE facts_unique_transform
+      (factset_id bigint NOT NULL,
+       fact_path text NOT NULL,
+       fact_value_hash varchar(40) NOT NULL)"
+
+   "INSERT INTO facts_unique_transform (factset_id, fact_path, fact_value_hash)
+      SELECT f.factset_id, fp.path, fv.value_hash
+        FROM facts f
+        INNER JOIN fact_values fv on f.fact_value_id = fv.id
+        INNER JOIN fact_paths fp on fv.path_id = fp.id"
+
+   "DROP TABLE facts"
+
+   ;; Remove all the orphaned duplicates (all but the row in each set
+   ;; with min-id).
+   (if (scf-utils/postgres?)
+     "DELETE FROM fact_paths t1 USING fact_paths t2
+        WHERE t1.path = t2.path AND t1.id > t2.id"
+     "DELETE FROM fact_paths t1
+        WHERE t1.id <> (SELECT MIN(t2.id) FROM fact_paths t2
+                          WHERE t1.path = t2.path)")
+   (if (scf-utils/postgres?)
+     "DELETE FROM fact_values t1 USING fact_values t2
+        WHERE t1.value_hash = t2.value_hash AND t1.id > t2.id"
+     "DELETE FROM fact_values t1
+        WHERE t1.id <> (SELECT MIN(t2.id) FROM fact_values t2
+                          WHERE t1.value_hash = t2.value_hash)")
```

Comment:
@wkalt well "t1.id > t2.id" was the query I wanted to use, but hsqldb couldn't handle it.